superset/daos/key_value.py

```diff
@@ -0,0 +1,195 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+from datetime import datetime
+from typing import Any
+from uuid import UUID
+
+from sqlalchemy import and_
+
+from superset import db
+from superset.daos.base import BaseDAO
+from superset.key_value.exceptions import (
+    KeyValueCreateFailedError,
+    KeyValueUpdateFailedError,
+)
+from superset.key_value.models import KeyValueEntry
+from superset.key_value.types import KeyValueCodec, KeyValueResource
+from superset.key_value.utils import get_filter
+from superset.utils.core import get_user_id
+
+logger = logging.getLogger(__name__)
+
+
+class KeyValueDAO(BaseDAO[KeyValueEntry]):
+    @staticmethod
+    def get_entry(
+        resource: KeyValueResource,
+        key: int | UUID,
+    ) -> KeyValueEntry | None:
+        """
+
+        :param resource:
+        :param key:
+        :return:
+        """
+        filter_ = get_filter(resource, key)
+        return db.session.query(KeyValueEntry).filter_by(**filter_).first()
+
+    @classmethod
+    def get_value(
+        cls,
+        resource: KeyValueResource,
+        key: int | UUID,
+        codec: KeyValueCodec,
+    ) -> Any:
+        """
+
+        :param resource:
+        :param key:
+        :param codec:
+        :return:
+        """
+        entry = cls.get_entry(resource, key)
+        if not entry:
+            return None
+
+        return codec.decode(entry.value)
+
+    @staticmethod
+    def delete_entry(resource: KeyValueResource, key: int | UUID) -> bool:
+        """
+        Delete an entry if it exists
+
+        :param resource: The resource
+        :param key:
+        :return: `True` if an entry was deleted, `False` if not
+        """
+        if entry := KeyValueDAO.get_entry(resource, key):
+            db.session.delete(entry)
+            return True
+
+        return False
+
+    @staticmethod
+    def delete_expired_entries(resource: KeyValueResource) -> None:
+        (
+            db.session.query(KeyValueEntry)
+            .filter(
+                and_(
+                    KeyValueEntry.resource == resource.value,
+                    KeyValueEntry.expires_on <= datetime.now(),
+                )
+            )
+            .delete()
+        )
+
+    @staticmethod
+    def create_entry(
+        resource: KeyValueResource,
+        value: Any,
+        codec: KeyValueCodec,
+        key: int | UUID | None = None,
+        expires_on: datetime | None = None,
+    ) -> KeyValueEntry:
+        """
+
+        :param resource:
```

Comment:
If you're not going to define these I would just remove the docstring.

Reply:
Oops, good catch, I meant to add docstrings, but later they felt redundant. I'll remove them.